rfcs/0006-transaction.md

```diff
@@ -0,0 +1,299 @@
+# SlateDB Transaction
+
+<!-- TOC start (generated with https://github.com/derlin/bitdowntoc) -->
+
+- [SlateDB Transaction](#slatedb-transaction)
+  - [Background](#background)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+  - [Constraints](#constraints)
+  - [References](#references)
+  - [Requirements](#requirements)
+    - [Snapshot](#snapshot)
+    - [Write Batch](#write-batch)
+    - [Isolation Levels](#isolation-levels)
+  - [Other Systems](#other-systems)
+    - [RocksDB](#rocksdb)
+    - [Badger](#badger)
+  - [Proposal](#proposal)
+    - [API](#api)
+    - [Conflict Checking](#conflict-checking)
+    - [How to Test](#how-to-test)
+  - [Roadmap](#roadmap)
+  - [Updates](#updates)
+
+<!-- TOC end -->
+
+Status: Reviewing
+
+Authors:
+
+* [Li Yazhou](https://github.com/flaneur2020)
+
+## Background
+
+Transaction allows multiple operations to be executed in a single atomic operation, and control the isolation level of the operations. This is a planned feature in SlateDB, and is essential for many critical use cases, like:
+
+- Using SlateDB as a metadata store or coordinator for a distributed system, such as the backend of [Kine](https://github.com/k3s-io/kine) or various metadata stores in OLAP systems.
+- Using SlateDB as a underlying KV layer in an OLTP system, such as a [TiKV](https://tikv.org/) replacement or other SQL engines.
+
+This RFC proposes the goals & design proposal of the transaction feature in SlateDB.
+
+## Goals
+
+- Define the Transaction API in SlateDB.
+- Discuss the expected isolation level semantics in SlateDB.
+- Organize the prerequisites needed to implement Transaction feature.
+- Organize the roadmap for the implementation of transaction feature.
+
+## Non-Goals
+
+- Pessimistic Transaction support: although it is a supported feature in rocksdb, but it is not a goal for the initial implementation. Optimistic transaction is more widely used in practice, and considered as more scalable and efficient. We'll not discuss the support of pessimistic transaction in this RFC.
+- Distributed Transaction support: SlateDB is a single-writer database, and distributed transaction is not a goal for the initial implementation. We may consider support [Two Phase Commit](https://github.com/facebook/rocksdb/wiki/Two-Phase-Commit-Implementation) in the future, which may help users coordinate between multiple SlateDB instances, but it is a separate topic and not discussed in this RFC.
```

Comment:
Do you already have use cases in mind for multiple DB instances? 
Should we capture this as an enhancement issue so that the context is tracked?

Reply:
It would be useful to be able to do a 2PC across SlateDB and other systems. For example, you could do a 2PC across a Kafka Transaction (support for external transaction coordinators is being added to Kafka: https://cwiki.apache.org/confluence/display/KAFKA/KIP-939%3A+Support+Participation+in+2PC) to get transactional processing of records in Kafka that uses SlateDB to store intermediate state. To do this we'd need to be able to durably store prepared transactions that are awaiting commit.

Reply:
Until such first class support is there, perhaps prepare/commit stages could still be persisted by application layer explicitly, for some sort of limited support

Reply:
rocksdb also have 2pc support: https://github.com/facebook/rocksdb/wiki/Two-Phase-Commit-Implementation

in my memory, this 2pc is useful on the coordination of binlog and table data.

but even without built-in 2pc, users can still make 2pc possible over KV + local transaction, like Percolator's approach.

i guess 2pc is a standalone big topic, we can put it in another RFC after transaction got implemented, wdyt?